src/treq/test/test_content.py

```diff
@@ -215,3 +218,45 @@ def test_text_content_unicode_headers(self):
         self.protocol.connectionLost(Failure(ResponseDone()))
 
         self.assertEqual(self.successResultOf(d), u'ᚠᚡ')
+
+
+class UnfinishedResponse(Resource):
+    """Write some data, but never finish."""
+
+    isLeaf = True
+
+    def render(self, request):
+        request.write(b"HELLO")
+        return NOT_DONE_YET
+
+
+class MoreRealisticContentTests(TestCase):
+    """Tests involving less mocking."""
+
+    def test_exception_handling(self):
+        """
+        An exception in the collector function:
+
+            1. Always gets returned in the result ``Deferred`` from
+               ``treq.collect()``.
+
+            2. Closes the transport.
+        """
+        stub = StubTreq(UnfinishedResponse())
+        response = self.successResultOf(stub.request("GET", "http://127.0.0.1/"))
+        self.assertEqual(response.code, 200)
+
+        def error(data):
+            1 / 0
+
+        d = collect(response, error)
+
+        # Exceptions in the collector are passed on to the caller via the
+        # response Deferred:
+        self.failureResultOf(d, ZeroDivisionError)
+
+        # An exception in the protocol results in the transport for the request
+        # being closed.
+        self.assertTrue(
+            response.original.original._transport._producer.disconnecting
```

Comment:
Wouldn't it be cleaner to have `UnfinishedResponse` capture that it was disconnected?

Reply:
OK, I figured out a way to do that.